identity/aziot-hub-client-async/src/lib.rs

```diff
@@ -141,328 +154,270 @@ impl Client {
+                        self.proxy_uri.clone(),
+                        false,
+                    )
+                    .await?;
+
+                    let authorization_header_value =
+                        hyper::header::HeaderValue::from_str(&token)
+                            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+                    req.headers_mut()
+                        .append(hyper::header::AUTHORIZATION, authorization_header_value);
+                    connector
                 }
-            }
-        }
-
-        let body = hyper::body::to_bytes(body)
-            .await
-            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
-
-        let res: TResponse = match res_status_code {
-            hyper::StatusCode::OK | hyper::StatusCode::CREATED => {
-                if !is_json {
-                    return Err(std::io::Error::new(
-                        std::io::ErrorKind::Other,
-                        "malformed HTTP response",
-                    ));
+                aziot_identity_common::Credentials::Tpm => {
+                    let audience = format!(
+                        "{}/devices/{}",
+                        self.device.iothub_hostname, self.device.device_id
+                    );
+                    let (connector, token) = get_sas_connector(
+                        &audience,
+                        (),
+                        &*self.tpm_client,
+                        self.proxy_uri.clone(),
+                        false,
+                    )
+                    .await?;
+
+                    let authorization_header_value =
+                        hyper::header::HeaderValue::from_str(&token)
+                            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+                    req.headers_mut()
+                        .append(hyper::header::AUTHORIZATION, authorization_header_value);
+                    connector
                 }
-                let res = serde_json::from_slice(&body)
-                    .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
-                res
-            }
+                aziot_identity_common::Credentials::X509 {
+                    identity_cert,
+                    identity_pk,
+                } => {
+                    get_x509_connector(
+                        &identity_cert,
+                        &identity_pk,
+                        &self.key_client,
+                        &mut *self.key_engine.lock().await,
+                        &self.cert_client,
+                        self.proxy_uri.clone(),
+                    )
+                    .await?
+                }
+            };
+
+            let client: hyper::Client<_, hyper::Body> = hyper::Client::builder().build(connector);
+
+            let err = match tokio::time::timeout(self.req_timeout, client.request(req)).await {
```

Comment:
Can you put this in `http-common` to reuse it? It's already duplicated between hub-client and dps-client, and in the future certd might need it too.

Reply:
I tried merging them at first, but these functions each do some Hub or DPS-specific things and have some HubClient/DpsClient-specific parameters (such as self). It would be more rework to merge these with the common code.